### ISSUE TITLE
MSK cluster destroy doesn't remove `aws_msk_configuration`

### ISSUE THREAD

**Issue by carlosperello** [terraform-providers/terraform-provider-aws#9082]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.10.8

### Affected Resource(s)

* aws_msk_configuration

### Terraform Configuration Files

```hcl
resource "aws_msk_cluster" "msk" {
  cluster_name           = "${var.cluster_name}"
  kafka_version          = "${var.kafka_version}"
  number_of_broker_nodes = "${var.number_of_broker_nodes}"

  broker_node_group_info {
    instance_type  = "${var.broker_instance_type}"
    ebs_volume_size = "${var.broker_ebs_volume_size}"
    client_subnets = [ "${data.terraform_remote_state.vpc.private_persistence_subnet_ids}" ]
    security_groups = [ "${aws_security_group.msk.id}" ]
  }

  configuration_info {
    arn      = "${aws_msk_configuration.msk.arn}"
    revision = "${aws_msk_configuration.msk.latest_revision}"
  }
}

resource "aws_msk_configuration" "msk" {
  kafka_versions = ["2.1.0"]
  name           = "msk"

  # https://docs.aws.amazon.com/msk/latest/developerguide/msk-configuration-properties.html
  server_properties = <<PROPERTIES
auto.create.topics.enable = true
delete.topic.enable = true
PROPERTIES
}

```

### Debug Output

Basically, `aws_msk_configuration` is not considered to be deleted.

### Panic Output
None

### Expected Behavior

`aws_msk_configuration` is destroyed together with the cluster definition.

### Actual Behavior

`aws_msk_configuration`is kept around, preventing from creating the cluster again:

´´´
Error: Error applying plan:

1 error(s) occurred:

* aws_msk_configuration.msk: 1 error(s) occurred:

* aws_msk_configuration.msk: error creating MSK Configuration: ConflictException: A resource with this name already exists.
	status code: 409, request id: 75c49562-9451-11e9-8669-6790fafadadb

´´´

### Steps to Reproduce

1. `terraform apply` with the definition of an msk cluster with an `aws_msk_configuration`
2. `terraform destroy`
3. `terraform apply` will fail.

### Important Factoids

None

### References

None

**Comment by bflad** [comment]
Hi @carlosperello 👋 

The [MSK API](https://docs.aws.amazon.com/msk/1.0/apireference/configurations-arn.html) does not currently provide an API call to delete MSK Configurations, so there is nothing we can do within Terraform for this particular issue. If you are interested in the ability to delete MSK Configurations, please create a feature request through an AWS Support Case or contact your AWS account team if you have one.

We try to call out this limitation in the [`aws_msk_configuration` resource](https://www.terraform.io/docs/providers/aws/r/msk_configuration.html):

> NOTE: The API does not support deleting MSK configurations. Removing this Terraform resource will only remove the Terraform state for it.

If you have any suggestions for how to improve this messaging, please let us know.

**Comment by carlosperello** [comment]
Hmm, didn't realize it was an AWS limitation, however, I think the bug still applies, it's impossible to destroy a cluster and then create it again unless you rename the configuration. If it cannot be removed, I think at least we should retrieve the existing one to verify it's the same content and use its ARN directly instead of failing due to an existing config with the same name instead of having to do it manually... Maybe I'm missing something obvious though...

Thanks.

**Comment by bflad** [comment]
Terraform as a whole does not make any assumptions about existing infrastructure and will only attempt to manage what you tell it to manage given its current view of your infrastructure ("state" in Terraform terminology). This design philosophy has been a part of Terraform since its inception and applies to practically every Terraform resource.

There are a few options here:

* You can explicitly bring existing resources under management by importing them, mentioned above with `terraform import`.
* You can place the `aws_msk_configuration` resource configuration outside where the `aws_msk_cluster` configuration is placed, so it is not destroyed/recreated when the `aws_msk_cluster` configuration is destroyed/recreated.
* We can potentially implement an `aws_msk_configuration` data source, so that configurations can have a validated, read-only view of the existing resource to help with the above.

We're not likely to make an exception to the state-based design philosophy for this resource as that would introduce confusing differences for how Terraform operates with the majority of other infrastructure, but if the data source option sounds like it might be helpful in your environment we can reopen this issue and retag it as a feature request.

**Comment by carlosperello** [comment]
Fair enough. I guess the third option is the best one with current AWS API state to avoid confusions about how to recover from current broken state we end having after a destroy.

Thank you.

**Comment by bflad** [comment]
Sorry, I missed two other options 😅 

* You can use the [Terraform Random Provider](https://www.terraform.io/docs/providers/random/index.html) to have your Terraform configuration automatically generate new names for the configurations.
* Similar to the above, we can implement a `name_prefix` argument to the resource and/or allow `name` to be completely optional and randomly generated.

We generally shy away from automatically implementing the `name_prefix` approach for resources unless there's a strong use case, but given the current situation with this resource, we certainly would consider it. That said, in talking with the MSK team directly, deleting configurations is on their radar so hopefully this is a stopgap issue.

I'll go ahead and reopen this and retitle it for the feature request. The MSK implementation is still in my recent context so I might just go ahead and submit the new data source right now so we have it available. 👍

**Comment by bflad** [comment]
Pull request submitted for new data source: #9088

**Comment by bflad** [comment]
The new `aws_msk_configuration` data source has been merged and will release with version 2.19.0 of the Terraform AWS Provider, likely later this week. 👍

**Comment by bflad** [comment]
The `aws_msk_configuration` data source has been released in [version 2.19.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2190-july-11-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!
